### PR TITLE
Add azure image module support create and delete

### LINKED ISSUE #30940: Azure module for Managed Images
It would be greate if a module for Azure could be writting to interact with Managed Images. Currently, Azure supports two methonds for interacting with images. 1. Images from the Marketplace that someone else builds and maintains. 2. Managed Images that I'm responsible for building and maintaining, and appling custom packages.

Could a module be written to interact with Managed Images? The CLI arguments start with "az image", and there's concepts of list/create/delete. 

This would potentially let us manage these images with tower.

##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
azurerm modules

### PR DESCRIPTION
**Pull request by yuwzho**
##### SUMMARY

Add Azure Image module support
It supports
- Create an image from an existing VM.
- Create an image from OS disk and data disk with OS type specific, where the disk can be a disk, blob uri or snapshot.
- Delete an existing image.

 Fix #30940

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

azure_rm_image.py

##### ANSIBLE VERSION

```
2.5
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by xscript** [comment]
shipit

**Comment by yaweiw** [comment]
shipit

**Comment by jborean93** [comment]
There is another PR to add this module https://github.com/ansible/ansible/pull/32403 and we may need to decide which one to move forward with. This implementation has the ability to source from another source type which makes me lean more towards it but we should decide before merging one of them.

**Comment by elaurijssens** [comment]
I'd vote for this one over #32403, but let's see if that PR's facts module can be adapted for use here. We do need a facts module to get information about images, it's essential wrt how to deploy them.

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: Blanket exception-swallowing like this is dangerous and could mask a number of other issues... Should probably just catch `CloudError` and check if `error.error` is `ResourceNotFound`- if it's not, `raise` and be done.

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: This validation should be included in the `check_mode` behavior - the usual philosophy there is to do as much as possible without actually making changes to both check for changes and ensure that it would work.

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: undocumented return value; should probably be removed or folded into a diff_mode impl

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: I think we've fixed the underlying bug that necessitates this update (the model/msrest serializer change around not serializing `readonly` model values), so this should no longer be necessary

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (nitzmahone) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: `required: false` is not necessary and can be removed from all these

**Review by nitzmahone — CHANGES_REQUESTED** [review]
Couple of minor things, but generally looks great! Agree that we should probably have a separate facts module for typical workflows, but we can work that out independently.

**Comment by jctanner** [comment]
bot_status

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: Agree with u, and I will change that

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: Good point, so I move the `check_mode` condition check at the last step before call the create/delete API

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: Actually the `self.serialize_obj` has not been fixed, it still missing these field. But `as_dict` is perfect

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: It should check the `CloudError.error.error` is `ResourceNotFound`. I check the `status_code == 404` instead

**Comment by yuwzho** [comment]
@nitzmahone I have make some changes, please take a look 😄

**Comment by yuwzho** [comment]
ready_for_review

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: Can we filter the values we return to only be things we don't know.

* The ID of the disk
* Potentially the underlying data disks of the image

These values should also be set on the root level and not in `state` so I could do

```
- name: create image
  azure_rm_image:
    resource_group: Testing
    name: foobar
    source: testvm001

  register: image_info

- name: show image id
  debug:
    var: image_info.id
```

The facts module is a different story but we usually only return values for things we don't know.

File: test/integration/targets/azure_rm_virtualmachine/tasks/virtualmachine.yml
Comment (jborean93) [inline test/integration/targets/azure_rm_virtualmachine/tasks/virtualmachine.yml]: Can we have 3 tests per operation

* test with check mode to make sure an image wasn't actually created
* test normally (what is here right now)
* test again to check idempotency and make sure no change is recorded

**Review by jborean93 — COMMENTED** [review]
Only 2 things I have are around the return values and the tests.

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: I m confused what should we return in the module. Is there any overhead for return everything I know to user? Consider the maintenance, it seems more convenient if we keep the struct as Azure SDK returned instead of flatting it and deleting some information in the module layer.

Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: There are a few reasons;

* if you are just converting the result from the dict  you are bound the following restrictions
    * any changes in this return value get's reflected to the module and if a playbook is reliant on these values they will now break, we want to keep things consistent
    * the return values are akin to being an API as well, we want to ensure we keep things as simple so that we don't accrue technical debt
    * the structure itself can be in camelCase format which we try to avoid in Ansible to keep consistency in the values
    * Sometimes what people look for is deeply nested in the return value and should be more accessible, case in point when trying to get the public IP for an instance created with `azure_rm_virtualmachine` I need to do `result.properties.networkProfile.networkInterfaces[0].properties.ipConfigurations[0].properties.publicIPAddress.properties.ipAddress`. Compare this to the EC2 module, to get the public IP I only need to do `result.instances[0].public_ip` which is a hell of a lot easier for a non programmer to derive and understand
* we usually advocate only sending info we don't know already (in this case the ID) as the return values are mostly used for further actions on the newly created object. If we wanted to get more information about an object, that's why we have facts modules.
* each byte being sent across the wire takes more time, while in this case it would make a marginal difference why would you need to send the name of an image when you already know that value

From a developer perspective it is simpler to return the Azure SDK result instead of flattening it but from an Ansible user perspective, it is more convenient for them to have a simple interface.

Ultimately the first point is the main reason, and it comes down to the fact that we want a nice and consistent interface that doesn't change over time. In a perfect world we could probably have that but unfortunately that is not this world so we keep things simple as we can.

**Comment by yuwzho** [comment]
ready_for_review

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: Thanks for you explain, I have changed the _image_to_dict from simply serialize it to flatten it.

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: Should probably state that if not specified, the location is derived from the resource group location

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: replace `'present'` with `C(present)` and `'absent'` with `C(absent)`

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: No worries, I see your changes but it just needs to be taken 1 step further. Right now the results are being added to `result['state']` when each value should just be in result, e.g. ID should be `result['id']` and not `result['state']['id']`

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: state should not be in the results object

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: We already know the name of the image and so don't need to return this

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: Already know, don't return

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: Already know, don't return. Also when setting return values, Ansible used snake_case and not camelCase.

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: s/osDisk/os_disk/

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: s/dataDisks/data_disks

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: Already know this, don't return.

File: test/integration/targets/azure_rm_image/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/azure_rm_image/tasks/main.yml]: Are you able to do the same with this one, run delete in check mode and then a normal delete again to ensure it reports no change

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: Even the name is ignored? Sounds weird

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: Location maybe the default as resource group, not already know exactly

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: I think this value should be return. If we assume these all things user already knows, we can return the image id only. But it is so weird

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: I see your example that some modules return properties in the result's child level. But all other azure modules return property values in the state level. It is so strange that the azure image module return thing not consistent with other azure modules. A user who uses Azure modules may confuse with this.

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: We know the name, we must set this value so what's the point in returning it.

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: Fair enough

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: I don't think it is weird, there are plenty of modules where they don't return values at all. `os_disk`, `data_disks` and after thinking about it more this one is fine to return as it returns the full path of the VM and not just the name. The only change I would make is to only return it if the user set the vm name as the source.

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: Sounds that we only need to return the created resource's id, and a changed value.

Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_image.py]: That's about it, facts modules are different.

**Comment by yuwzho** [comment]
ready_for_review

**Review by jborean93 — APPROVED** [review]
Looking great!

**Comment by jborean93** [comment]
Are you able to raise a separate PR to add this as a new module to the changelog for 2.5.

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jeking3) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py]: This does not actually work if you follow the example.  If you place a bare list of names under data_disk_sources, this fails.  The documentation or the code is wrong.  When I add a dict containing lun and source, errors occur further on down.

File: lib/ansible/modules/cloud/azure/azure_rm_image.py
Comment (jeking3) [inline lib/ansible/modules/cloud/azure/azure_rm_image.py:250]: The lun parameter here needs to be named, otherwise the call fails.
I submitted a fix: #49394

**Pull request merged**
